### PR TITLE
#94 improve error message 

### PR DESCRIPTION
**Pull request by FrancoisPgm**
Add a NotFoundConfoundException class and use it to collect all the unfound params and raise a final error with the list of all not found params rather than just the first one.

### CODE REVIEW

**Comment by FrancoisPgm** [comment]
Ok I realised I didn't get the codeclimate issue, I saw refactoring and thought it was just some styling issue, so I just removed some empty lines that didn't seem to respect PEP8. 
But actually the issue is about _cognitive complexity_ of the `_load_single` method. Looking at their definition of _cognitive complexity_ it seems to be related to all the try and catch that I added, but I don't see a more explicit way to handle it. We could return the not found params arrays all the way back from the `_check_params` method but that seems way more clunky to me, whereas exceptions are made just for this kind of situations. 

A way of making the method shorter (but not necessarily more readable) would be to wrap all the 
```
if "<strat>" in self.strategy:
            try:
                confounds_<strat> = _load_<strat>(confounds_raw, self.<strat>)
                confounds = pd.concat([confounds, confounds_<strat>], axis=1)
            except ConfoundNotFoundException as exception:
                not_found_confounds += exception.params
```
blocks in a for loop. In my opinion that would be better because it would remove the duplicate code of the try/catch and concat but codeclimate might not like it since it would be less explicit.

File: load_confounds/parser.py [hunk 1]
```diff
@@ -340,30 +355,51 @@ def _load_single(self, confounds_raw):
...
+            try:
+                confounds_global_signal = _load_global(confounds_raw, self.global_signal)
+                confounds = pd.concat([confounds, confounds_global_signal], axis=1)
+            except ConfoundNotFoundException as exception:
+                not_found_confounds += exception.params
 
         if "compcor" in self.strategy:
-            confounds_compcor = _load_compcor(
-                confounds_raw, self.compcor, self.n_compcor
+            try :
```
Comment (pbellec) [inline load_confounds/parser.py:390]: Getting all those `try` `except` argument in the code is not elegant. As they are all written around private functions, maybe we could simply add decorators. Something like: 
```
def trace_error(func):
    def wrapper_trace_error(*args, **kwargs):
        try:
            func(*args, **kwargs)
        except ConfoundNotFoundException as exception:
            not_found_confounds += exception.params
        return func(*args, **kwargs)
    return wrapper_trace_error
``` 
Then all you would need to do is add this decorator before all of the `_load_xxx` functions. 
```
@trace_error
def _load_global( ...
``` 
See https://realpython.com/primer-on-python-decorators/ 
Then the code of `_load_single` would be unchanged, except for the last statement `if not_found_confounds:` wich could be turned into a short helper function.

**Review by pbellec — CHANGES_REQUESTED** [review]
Great PR @FrancoisPgm thanks for looking into this! The main thing to add for this to be merged is a test (or a series of tests) checking that the correct error messages are being generated when a subset of confounds is missing. I have also made a suggestion about using a decorator to make the `try ... except` blocks less repetitive.

**Comment by pbellec** [comment]
> A way of making the method shorter (but not necessarily more readable) would be to wrap all the
> blocks in a for loop. 

The reason why it's not a long for loop is because the arguments of the `_load_xxx` vary depending on the function. I guess it would be possible to add all the `_load_xxx` functions as attributes to the Confounds class, and use `self` as an argument and then access all the parameters as attributes of Confouds. Then there could be a loop over all elements of the global variable `all_confounds`.  

I have made a suggestion for using a decorator which, if it works, will solve the codeclimate issue. 
If you want to try to refactor the code in a for loop as well, go for it. Otherwise I'll try as a separate issue.
It will become handy when we add `ica-aroma` and `censoring` to the list of methods, to keep the code concise.

**Comment by FrancoisPgm** [comment]
Yay, all good !

**Comment by FrancoisPgm** [comment]
The only change I'm not totally sure about is to put `_load_high_pass` as a function and not a method of the `Confounds` class. It was required by codacy because `_load_high_pass` didn't use `self`. However it felt better to put all the load functions in the same place. 
It feels even better to have the "All checks have passed" icon though.

**Review by pbellec — APPROVED** [review]

**Comment by pbellec** [comment]
@FrancoisPgm I think this is ready to merge, but it includes a major refactoring. So I think I am going to try and merge PRs #99 #96 and #100 before this one. This means you  will need to merge the other PRs in your fork, fix the conflicts and refactor the new bits on ICA. Is that OK?

**Comment by FrancoisPgm** [comment]
Ok I'll try to do that.

**Comment by pbellec** [comment]
@FrancoisPgm your PR is finally ready to merge. 

Tell me if you need some help with the conflicts, I am also happy to take over and deal with the merge myself.

**Comment by pbellec** [comment]
So.... this merge implied some pretty big refactoring. This was mostly with the parts of the code that had to do with the other PR and the added support for censoring, ica and improved compcor. The main notable change is that I had to change completely the tests. I created by hand a test file with missing params and keywords. I put a single test for both things missing at the same time. The reason why I had to do this is that I dropped support for data frames. For compcor we need the json file, so it's not possible to directly transform a dataframe. 

Regarding the warnings: I manually disabled the codacy issues. It wants `_load_ica` and another one as functions rather than methods of the `Confounds` class. But there is a neat pattern where all the `_load_{confound}` method are exposed in the class. I really like that, and I don't care it could be a function. 

For the codecov warning, the lack of coverage relates to the censoring method, which is not currently tested.

**Comment by pbellec** [comment]
It's done, merging in main!! 

Congrats and huge thanks @FrancoisPgm I love the new class structure. 

@all-contributors add @FrancoisPgm for test and data

**Pull request merged**